scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -1665,3 +1668,420 @@ cdef FibonacciNode* remove_min(FibonacciHeap* heap) noexcept:
 #        print_node(heap.min_node)
 #    else:
 #        print("[empty heap]")
+
+######################################################################
+
+# Author: Tomer Sery  -- <tomersery28@gmail.com>
+# License: BSD, (C) 2024
+
+def yen(
+    csgraph,
+    source,
+    sink,
+    K,
+    directed=True,
+    return_predecessors=False,
+    unweighted=False,
+):
+    """
+    yen(csgraph, source, sink, K, directed=True, return_predecessors=False,
+        unweighted=False)
+    Yen's K-Shortest Paths algorithm on a directed or undirected graph.
+
+    Parameters
+    ----------
+    csgraph : array, matrix, or sparse matrix, 2 dimensions
+        The N x N array of distances representing the input graph.
+    source : int
+        The index of starting node for the paths.
+    sink : int
+        The index of the ending node for the paths.
+    K : int
+        The number of shortest paths to find.
+    directed : bool, optional
+        If True (default), then find the shortest path on a directed graph:
+        only move from point i to point j along paths csgraph[i, j].
+        If False, then find the shortest path on an undirected graph: the
+        algorithm can progress from point i to j along csgraph[i, j] or
+        csgraph[j, i]
+    return_predecessors : bool, optional
+        If True, return the size (N, N) predecessor matrix.
```

Comment:
These mutating signatures tend to be quite nasty especially as it doesn't save on computation. Elsewhere in SciPy we often use results objects, not sure if there is a president for this in sparse?

Reply:
I was going for the common API in `_shortest_paths`.
A results object is cleaner. Shall I change?

Reply:
That's a fair point, I'll leave that for a sparse resident to decide

Reply:
I think having a consistent API is better, but maybe let's raise a discussion on the mailing list about moving the csgraph API to be a return object.

Reply:
Unresolving, since this set of comments is referred to on the mailing list and otherwise hard to find.